### PR TITLE
add interface and mock

### PR DESCRIPTION
**Pull request by Igor-Rashyn**

### CODE REVIEW

File: redis/config.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package redis
```
Comment (EugenRakhimov) [inline redis/config.go:1]: I would suggest to make it a separate package, and put it inside subdirectory, something like redis/config/config.go

File: redis/cache.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	Get(key string) (string, error)
+	Exists(key string) (bool, error)
+	Set(key, val string) error
+	SetExpiry(key, val string, expireTime int) error
+	SetExpiryTime(key, val string, expireTime time.Time) error
+	Subscribe(subscription string, handleResponse func(interface{}))
+}
+
+type cache struct {
+	Pool *redisc.Cluster
```
Comment (EugenRakhimov) [inline redis/cache.go:29]: make it local, there is no need for it to be accessible outside

File: redis/cache.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	cfg  *Config
+}
+
+// Conn returns an active connection to the cache
+func (c cache) Conn() redis.Conn {
+	return c.Pool.Get()
+}
+
+func addr(cfg *Config) string {
```
Comment (EugenRakhimov) [inline redis/cache.go:38]: it should be a function on Config struct,  rather the function like this. In fact - Config should be the struct that returns Address

File: redis/cache.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	return fmt.Sprintf("%s:%s", cfg.Host, cfg.Port)
+}
+
+func NewCache(cfg *Config) (*cache, error) {
+	redisPool := &redisc.Cluster{
+		CreatePool:   createPool,
+		StartupNodes: []string{addr(cfg)},
```
Comment (EugenRakhimov) [inline redis/cache.go:45]: instead it should be cfg.Address()

File: redis/config.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"github.com/HomesNZ/go-common/env"
+	validation "github.com/go-ozzo/ozzo-validation"
+)
+
+type DeclareConfig func(config *Config)
+
+type Config struct {
```
Comment (EugenRakhimov) [inline redis/config.go:10]: make it interface as well, and no need to expose Host and Port, you can make the address function instead.

File: redis/config.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	}
+
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+func (c Config) Validate() error {
```
Comment (EugenRakhimov) [inline redis/config.go:31]: No need to expose this function outside of this package.

File: redis/config.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+type DeclareConfig func(config *Config)
+
+type Config struct {
+	Host string
+	Port string
+}
+
+func ConfigFromEnv() (*Config, error) {
```
Comment (EugenRakhimov) [inline redis/config.go:15]: Would be good to add test ;)

**Review by EugenRakhimov — APPROVED** [review]
Overall looks good, though added some comment about better structuring

File: redis/cache.go (hunk 2 above)
Comment (Igor-Rashyn) [reply redis/cache.go:29]: thanks

File: redis/cache.go (hunk 3 above)
Comment (Igor-Rashyn) [reply redis/cache.go:38]: thanks

File: redis/cache.go (hunk 4 above)
Comment (Igor-Rashyn) [reply redis/cache.go:45]: thanks

**Pull request merged**